compiler/codegen/CodeGenPrep.cpp

```diff
@@ -177,6 +179,125 @@ OMR::CodeGenerator::lowerTreesPreChildrenVisit(TR::Node * parent, TR::TreeTop *
+ *
+ * The `newvalue` opcode is primarely useful to the optimizer for
+ * representing fused allocation an initialization. However, it's
+ * functionality can also be implemented using the `new` opcode
+ * (although doing so makes analyses more difficult in the optimizer).
+ *
+ * This function takes a `newvalue` tree and lowers it to an equivalent
+ * sequence of trees using a `new` opcode. For example, the following
+ * tree:
+ *
+ *    treetop
+ *      newvalue jitNewValueObject (identityless)
+ *        loadaddr Vec3
+ *        fload x
+ *        fload y
+ *        fload z
+ *
+ * will turned into:
+ *
+ *    treetop
+ *      fload x
+ *    treetop
+ *      fload y
+ *    treetop
+ *      fload z
+ *    treetop
+ *      new jitNewValueObject (skipZeroInit)
+ *        loadaddr Vec3
+ *    istorei Vec3.x
+ *      ==>new
+ *      ==>fload x
+ *    istorei Vec3.y
+ *      ==>new
+ *      ==>fload y
+ *    istorei Vec3.z
+ *      ==>new
+ *      ==>fload z
+ *    fence
+ *
+ * Note that the (helper) symbol reference for the `new` node is be
+ * the same as the symbol reference for the `newvalue` node.
+ *
+ * The fence at the end is needed for platforms with weak memory ordering
+ * such as POWER. It may appear anywhere after allocation + initialization
+ * but before any operation that could publish the resulting reference to
+ * other threads.
+ *
+ * @param comp pointer to the compilation object
+ * @param node the node being lowered
+ * @param tt the TreeTop anchoring the node
+ */
+static void
+lowerNewValue(TR::Compilation *comp, TR::Node *node, TR::TreeTop *tt)
+   {
+   auto* prevTreeTop = tt->getPrevTreeTop();
+
+   // remove tt from the TreeTop List
+   prevTreeTop->setNextTreeTop(NULL);
+
+   // Iterate over every child that sets a field. The first
+   // child is skipped as it's only used to specify the type
+   // of the value being constructed. The second child is
+   // the one that actually initializes the first field.
+   for (int i = 1; i < node->getNumChildren(); ++i)
+      {
+      auto* ttNode = TR::Node::create(TR::treetop, 1);
+      ttNode->setFirst(node->getChild(i));
+      prevTreeTop = TR::TreeTop::create(comp, ttNode, NULL, prevTreeTop);
+      prevTreeTop->getPrevTreeTop()->setNextTreeTop(prevTreeTop);
+      }
+
+   // Create new TreeTop to anchor the new. The TreeTop for
+   // the newValue cannot be reused because it is still used
+   // just after lowering the node.
+   auto* newNode = TR::Node::createWithSymRef(TR::New, 1, node->getSymbolReference());
+   newNode->setFirst(node->getChild(0));
+   auto* ttNode = TR::Node::create(TR::treetop, 1);
+   ttNode->setAndIncChild(0, newNode);
+   prevTreeTop = TR::TreeTop::create(comp, ttNode, NULL, prevTreeTop);
+   prevTreeTop->getPrevTreeTop()->setNextTreeTop(prevTreeTop);
+
+   // Get the class and layout of the value type
+   auto* valueClass = static_cast<TR_OpaqueClassBlock *>(newNode->getSymbol()->getStaticSymbol()->getStaticAddress());
```

Comment:
Sorry this comment is coming in late.  In playing with this change, I just noticed that this line seems to be getting the symbol for the `jitNewValue` helper method rather than the class for the value type.

Should this be as follows, to get the symbol from the `loadaddr`?

```
auto* valueClass = static_cast<TR_OpaqueClassBlock *>(newNode->getChild(0)->getSymbol()->getStaticSymbol()->getStaticAddress());
```

Reply:
Yes, that's correct.